feature-staking-impl/src/main/java/io/novafoundation/nova/feature_staking_impl/data/validators/KnownNovaValidators.kt

```diff
@@ -1,41 +1,61 @@
 package io.novafoundation.nova.feature_staking_impl.data.validators
 
-import io.novafoundation.nova.common.utils.toHexAccountId
-import io.novafoundation.nova.runtime.ext.Geneses
+import io.novafoundation.nova.common.utils.filterNotNull
+import io.novafoundation.nova.runtime.ext.accountIdOf
+import io.novafoundation.nova.runtime.multiNetwork.ChainRegistry
 import io.novafoundation.nova.runtime.multiNetwork.chain.model.Chain
 import io.novafoundation.nova.runtime.multiNetwork.chain.model.ChainId
+import io.novafoundation.nova.runtime.multiNetwork.chainsById
+import io.novasama.substrate_sdk_android.extensions.toHexString
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
 
 interface KnownNovaValidators {
 
-    fun getValidatorIds(chainId: ChainId): Set<String>
+    suspend fun getValidatorIds(chainId: ChainId): List<String>
 }
 
-class FixedKnownNovaValidators : KnownNovaValidators {
+class RemoteKnownNovaValidators(
+    private val validatorsApi: NovaValidatorsApi,
+    private val chainRegistry: ChainRegistry,
+) : KnownNovaValidators {
 
-    private val novaValidators by lazy {
-        val sharedAccounts = sharedValidatorsAccountIdsHex()
+    private var validatorsByNetwork: Map<String, List<String>>? = null
+    private val validatorsMutex = Mutex()
 
-        mapOf(
-            Chain.Geneses.POLKADOT to sharedAccounts,
-            Chain.Geneses.KUSAMA to kusamaValidators(),
-            Chain.Geneses.ALEPH_ZERO to sharedAccounts
-        )
+    override suspend fun getValidatorIds(chainId: ChainId): List<String> {
+        return getValidators()[chainId].orEmpty()
     }
 
-    override fun getValidatorIds(chainId: ChainId): Set<String> {
-        return novaValidators[chainId].orEmpty()
+    private suspend fun getValidators(): Map<String, List<String>> {
+        return validatorsMutex.withLock {
+            if (validatorsByNetwork == null) {
+                validatorsByNetwork = fetchValidators()
```

Comment:
Probably, we don't want to cache empty list in case request is failed

Reply:
Thanks for spotting this. I have concern that not caching failed result might lead potential inconsistencies in recommendations within one session. But your suggestion is valid, I will think how can we improve this